src/Analysis/Codelyzer.Analysis.Common/MSBuildDetector.cs

```diff
@@ -0,0 +1,294 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.Setup.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace Codelyzer.Analysis.Common
+{
+    /// <summary>
+    /// Detects the MSBuild Path that best matches the solution
+    /// </summary>
+    public class MSBuildDetector
+    {
+        private List<VisualStudioInstanceData> _visualStudioInstances;
+
+        private const string SolutionFileVSVersion = "VisualStudioVersion";
+
+        public class VisualStudioInstanceData
+        {
+            public VisualStudioInstanceData()
+            {
+                Workloads = new List<string>();
+            }
+            public string Id { get; set; }
+            public string Name { get; set; }
+            public string Version { get; set; }
+            public string InstallationPath { get; set; }
+            public List<string> Workloads { get; set; }
+            public string VisualStudioMSBuildPath { get; set; }
+        }
+
+        public MSBuildDetector()
+        {
+            _visualStudioInstances = GetVisualStudioInstallations();
+        }
+
+        public List<VisualStudioInstanceData> GetVisualStudioInstanceData() => _visualStudioInstances;
+
+        public string MatchSolution(string solutionPath)
+        {
+            var version = GetVisualStudioVersionNumber(solutionPath);
+
+            var solutionMajorVersion = GetMajorVersion(version);
+
+            var firstInstanceWithMatchingVersion = _visualStudioInstances?.OrderByDescending(v => v.Version).FirstOrDefault(vr => solutionMajorVersion == GetMajorVersion(vr.Version));
+
+            //If there's a visual studio solution that matches the one used to create the solution:
+            if (firstInstanceWithMatchingVersion != null)
+            {
+                return firstInstanceWithMatchingVersion.VisualStudioMSBuildPath;
+            }
+            //If no version matches, return the max version
+            else
+            {
+                return _visualStudioInstances?.OrderByDescending(v => v.Version).FirstOrDefault().VisualStudioMSBuildPath;
+            }
+        }
+
+        private int GetMajorVersion(string version)
+        {
+            int index = version.IndexOf('.');
+            if(index < 0)
+            {
+                index = version.Length;
+            }
+            return int.Parse(version.Substring(0, index));
+        }
+
+        private string GetVisualStudioVersionNumber(string solutionPath)
+        {
+            var solutionFileText = File.ReadAllText(solutionPath);
+            var lines = solutionFileText.Split(Environment.NewLine);
+            var constantLine = lines.FirstOrDefault(l => l.StartsWith(SolutionFileVSVersion));
```

Comment:
In case of a corrupted sln file, I assume we never reach this point, or will there be a situation where this is not found in the sln file

Reply:
This is always added to a solution file. If it's not there, then the code after it won't be needed, which means we'll exit. There is a try/catch for the whole logic that will log the error, and revert back to using the default msbuild.